plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -141,38 +183,60 @@ This might mean reinventing GHC's OccAnal...
 -}
 
 processTerm
-    :: Inlining tyname name uni fun a m
+    :: forall tyname name uni fun a m. Inlining tyname name uni fun a m
     => Term tyname name uni fun a
     -> m (Term tyname name uni fun a)
-processTerm = \case
-    v@(Var _ n) -> do
+processTerm = substFreshenedTerm <=< traverseOf termSubtypes substFreshenedType where
+    substFreshenedTerm :: Term tyname name uni fun a -> m (Term tyname name uni fun a)
```

Comment:
Right, I think I didn't communicate what I was thinking here well. This version is a bit of a misnomer, since for terms this function actually does all the inlining! What I was thinking of was something like this:
```
substName :: name -> m (Term tyname name uni fun a)
substTyName :: tyname -> m (Type tyname name uni fun a)
applyTypeSubstitution :: Type ... -> m (Type ...)
applyTypeSubstitution = typeSubstTyNamesM substTyName
processTerm :: Term ... -> m (Term ...)
```
`substName` and `substTyName` are similar, `processTerm` and `applyTypeSubstitution` are not, since the former is both applying a substitution and extending it.